File: realsense/enhanced_photography/depth_photo_object.cc
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "realsense/enhanced_photography/depth_photo_object.h"

#include <string>

#include "base/bind.h"
#include "base/logging.h"

namespace realsense {
namespace enhanced_photography {

DepthPhotoObject::DepthPhotoObject(PXCPhoto* photo)
    : photo_(photo) {
  handler_.Register("getColorImage",
                    base::Bind(&DepthPhotoObject::OnGetColorImage,
                               base::Unretained(this)));
  handler_.Register("getDepthImage",
                    base::Bind(&DepthPhotoObject::OnGetDepthImage,
                               base::Unretained(this)));
}

DepthPhotoObject::~DepthPhotoObject() {
  DestroyPhoto();
}

void DepthPhotoObject::DestroyPhoto() {
  if (photo_) {
    photo_->Release();
    photo_ = nullptr;
  }
}

void DepthPhotoObject::OnGetColorImage(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Image img;
  if (!photo_) {
    info->PostResult(GetColorImage::Results::Create(img,
        "Invalid photo object"));
    return;
  }

  PXCImage* imColor = photo_->QueryReferenceImage();
  if (!CopyColorImage(imColor, &img)) {
    info->PostResult(GetColorImage::Results::Create(img,
        "Failed to get color image data."));
    return;
  }

  info->PostResult(GetColorImage::Results::Create(img, std::string()));
}

void DepthPhotoObject::OnGetDepthImage(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Image img;
  if (!photo_) {
    info->PostResult(GetDepthImage::Results::Create(img,
        "Invalid photo object."));
    return;
  }

  PXCImage* imDepth = photo_->QueryDepthImage();
  if (!CopyDepthImage(imDepth, &img)) {
    info->PostResult(GetDepthImage::Results::Create(img,
        "Failed to get depth image data."));
    return;
  }

  info->PostResult(GetDepthImage::Results::Create(img, std::string()));
}

bool DepthPhotoObject::CopyColorImage(PXCImage* pxcimage,
                                      jsapi::depth_photo::Image* img) {
  if (!pxcimage) return false;

  PXCImage::ImageInfo image_info = pxcimage->QueryInfo();
  PXCImage::ImageData image_data;
  if (pxcimage->AcquireAccess(PXCImage::ACCESS_READ,
      PXCImage::PIXEL_FORMAT_RGB32, &image_data) < PXC_STATUS_NO_ERROR) {
    return false;
  }

  img->format = PixelFormat::PIXEL_FORMAT_RGB32;
  img->width = image_info.width;
  img->height = image_info.height;

  uint8_t* rgb32 = reinterpret_cast<uint8_t*>(image_data.planes[0]);
  for (int y = 0; y < image_info.height; y++) {
    for (int x = 0; x < image_info.width; x++) {
      int i = x * 4 + image_data.pitches[0] * y;
      img->data.push_back(rgb32[i + 2]);
      img->data.push_back(rgb32[i + 1]);
      img->data.push_back(rgb32[i]);
      img->data.push_back(rgb32[i + 3]);
    }
  }

  pxcimage->ReleaseAccess(&image_data);
  return true;
}

bool DepthPhotoObject::CopyDepthImage(PXCImage* depth,
                                      jsapi::depth_photo::Image* img) {
  if (!depth) return false;

  PXCImage::ImageInfo depth_info = depth->QueryInfo();
  img->format = PixelFormat::PIXEL_FORMAT_DEPTH;
  img->width = depth_info.width;
  img->height = depth_info.height;

  PXCImage::ImageData depth_data;
  if (depth->AcquireAccess(PXCImage::ACCESS_READ,
      PXCImage::PIXEL_FORMAT_DEPTH, &depth_data) < PXC_STATUS_NO_ERROR) {
    return false;
  }

  for (int y = 0; y < depth_info.height; ++y) {
    for (int x = 0; x < depth_info.width; ++x) {
      uint16_t* depth16 = reinterpret_cast<uint16_t*>(
          depth_data.planes[0] + depth_data.pitches[0] * y);
      img->data.push_back(depth16[x]);
    }
  }
  depth->ReleaseAccess(&depth_data);
  return true;
}

}  // namespace enhanced_photography
}  // namespace realsense

File: realsense/enhanced_photography/android/src/org/xwalk/extensions/realsense/enhancedphotography/DepthPhotoObject.java
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.xwalk.extensions.realsense.enhancedphotography;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.Byte;
import java.lang.Integer;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import com.intel.camera.toolkit.depth.photography.core.*;
import com.intel.camera.toolkit.depth.photography.reader.DepthReader;

import org.xwalk.extensions.common.*;

public class DepthPhotoObject extends BindingObject {
    private static final String TAG = "DepthPhotoObject";
    private DepthPhoto mDepthPhoto;
    private static final int bytesPerInt = Integer.SIZE / Byte.SIZE;

    public DepthPhotoObject() {
        mDepthPhoto = new DepthPhoto(new DepthContext());
        mHandler.register("loadXDM", this);
        mHandler.register("queryReferenceImage", this);
        mHandler.register("queryOriginalImage", this);
        mHandler.register("queryDepthImage", this);
        mHandler.register("queryRawDepthImage", this);
    }

    private void reportMessage(FunctionInfo info, String data, String error) {
        try {
            JSONArray result = new JSONArray();
            result.put(0, data);
            result.put(1, error);
            info.postResult(result);
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    private void queryColorImage(FunctionInfo info, Image image) {
        if (image == null) {
            reportMessage(info, "", "There is no image info");
            return;
        }

        PixelData colorData = image.getPixelData();
        Log.d(TAG, "PixelData: " + colorData.getWidth() + ", " + colorData.getHeight() +
                ", " + colorData.getPixelFormat() + ", " + colorData.getBytesPerPixel());
        int height = (int)colorData.getHeight();
        int width =  (int)colorData.getWidth();
        int pixelLength = width * height;
        // RGBA format is 4 bytes per pixel
        int pixelBytesLength = pixelLength * 4;
        byte[] colorBytes;

        if (colorData.getPixelFormat() == PixelData.PixelFormat.BGR) {
            byte[] srcBytes = colorData.copyByteDataOut();
            colorBytes = new byte[pixelBytesLength];

            for (int a = 0; a < srcBytes.length / 3; a++) {
                colorBytes[a * 4] = srcBytes[a * 3 + 2];      // b
                colorBytes[a * 4 + 1] = srcBytes[a * 3 + 1];  // g
                colorBytes[a * 4 + 2] = srcBytes[a * 3 + 0];  // r
                colorBytes[a * 4 + 3] = (byte)0xff;           // a
            }
        } else if (colorData.getPixelFormat() == PixelData.PixelFormat.RGBA) {
            byte[] srcBytes = colorData.copyByteDataOut();
            colorBytes = new byte[pixelBytesLength];

            for (int a = 0; a < srcBytes.length / 4; a++) {
                colorBytes[a * 4] = srcBytes[a * 4 + 2];
                colorBytes[a * 4 + 1] = srcBytes[a * 4 + 1];
                colorBytes[a * 4 + 2] = srcBytes[a * 4 + 0];
                colorBytes[a * 4 + 3] = srcBytes[a * 4 + 3];
            }
        } else if (colorData.getPixelFormat() == PixelData.PixelFormat.BGRA) {
            colorBytes = colorData.copyByteDataOut();
        } else {
            // FIXME(Jiajia): Need to support other PixelFormat
            reportMessage(info, "", "Unsupported pixel format");
            return;
        }

        // Binary message: callbackId(int), width(int), height(int), colorImageData
        ByteBuffer message = ByteBuffer.allocate((int) (3 * bytesPerInt + pixelBytesLength));
        message.order(ByteOrder.LITTLE_ENDIAN);
        message.rewind();
        message.putInt(Integer.parseInt(info.getCallbackId()));
        message.putInt(width);
        message.putInt(height);
        for (int i = 0; i < pixelLength; ++i) {
            message.put(colorBytes[i * 4 + 0]);
            message.put(colorBytes[i * 4 + 1]);
            message.put(colorBytes[i * 4 + 2]);
            message.put(colorBytes[i * 4 + 3]);
        }
        info.postResult(message.array());
    }

    public DepthPhoto getDepthPhoto() {
        return mDepthPhoto;
    }

    public void setDepthPhoto(DepthPhoto depthPhoto) {
        mDepthPhoto = depthPhoto;
    }

    public void onLoadXDM(FunctionInfo info) {
        try {
            ByteBuffer buffer = info.getBinaryArgs();
            int offset = buffer.position();
            int count = buffer.array().length - offset;
            File file = File.createTempFile("temp", ".jpg");
            String fileName = file.getAbsolutePath();
            Log.d(TAG, "fileName = " + fileName);
            FileOutputStream fo = new FileOutputStream(file);
            fo.write(buffer.array(), offset, count);
            fo.flush();
            fo.close();

            if (!DepthReader.isXDMFile(new DepthContext(), fileName)) {
                String errMsg = "The selected file is not a compatible image for this app.";
                reportMessage(info, "", errMsg);
                return;
            }

            DepthContext context = new DepthContext();
            mDepthPhoto = DepthReader.read(context, fileName);
            if (mDepthPhoto == null) {
                String errMsg = "loadXDM failed";
                reportMessage(info, "", errMsg);
                return;
            }
            reportMessage(info, "success", "");
        } catch (IOException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onQueryReferenceImage(FunctionInfo info) {
        if (mDepthPhoto == null) {
            String errMsg = "DepthPhoto is uninitialized.";
            reportMessage(info, "", errMsg);
            return;
        }

        Image image = mDepthPhoto.getPrimaryImage();
        queryColorImage(info, image);
    }

    public void onQueryOriginalImage(FunctionInfo info) {
        if (mDepthPhoto == null) {
            String errMsg = "DepthPhoto is uninitialized.";
            reportMessage(info, "", errMsg);
            return;
        }

        Image image = mDepthPhoto.getUneditedPrimaryImage();
        queryColorImage(info, image);
    }

    public void onQueryDepthImage(FunctionInfo info) {
        if (mDepthPhoto == null) {
            String errMsg = "DepthPhoto is uninitialized.";
            reportMessage(info, "", errMsg);
            return;
        }

        PixelData depthData = mDepthPhoto.getDepthMap().getDepthData();
        Log.d(TAG, "depthData: " + depthData.getWidth() + ", " + depthData.getHeight() +
                ", " + depthData.getPixelFormat() + ", " + depthData.getBytesPerPixel());

        byte[] depthBytes = depthData.copyByteDataOut();
        long byteLength =
                depthData.getHeight() * depthData.getWidth() * depthData.getBytesPerPixel();
        // Binary message: callbackId(int), width(int), height(int), depthImageData
        ByteBuffer message = ByteBuffer.allocate((int) (3 * bytesPerInt + byteLength));
        message.order(ByteOrder.LITTLE_ENDIAN);
        message.rewind();
        message.putInt(Integer.parseInt(info.getCallbackId()));
        message.putInt((int)depthData.getWidth());
        message.putInt((int)depthData.getHeight());
        for (int i = 0; i < byteLength; ++i) {
            message.put(depthBytes[i]);
        }
        info.postResult(message.array());
    }

    public void onQueryRawDepthImage(FunctionInfo info) {
        // FIXME(Jiajia): See https://github.com/otcshare/realsense-extensions-crosswalk/issues/204
        onQueryDepthImage(info);
    }
}
